### PR TITLE
enshrined paymaster snip + typo fixes to snip-16

### PR DESCRIPTION
**Pull request by iliav-starkware**

### CODE REVIEW

File: SNIPS/snip-20.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+1. Balance checks:
+    * User token balance should cover their bid given their specified max rate $$\mathtt{balance(user)}\geq \mathtt{max\_ amount}\times r(\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$ If success, CONTINUE; else, REJECT.
+    * Paymaster STRK balance should cover user bid $$\mathtt{balance(paymaster)}\geq \mathtt{max\_ amount}\times (\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$ If success, CONTINUE; else, REJECT.
+2. Run account `__validate__`. If success CONTINUE, else REJECT.
+3. Run paymaster `__validate_paymaster__`. If success, CONTINUE; else, REVERT and skip to fee invocations below.
+4. Run account `__execute__`. If success, CONTINUE; else, REVERT and skip to fee invocations below.
+5. Fee invocations:
+    * If failure was in `__validate_paymaster__`, user pays sequencer in designated fee token:
+        * Burn $\mathtt{used\_ amount}\times r\cdot\mathtt{base\_ price}$.
+        * Transfer $\mathtt{used\_ amount}\times r\cdot\mathtt{tip}$.
```
Comment (ilyalesokhin-starkware) [inline SNIPS/snip-20.md:78]: Does the user pay anything when r == 0?

Comment (leo-starkware) [reply SNIPS/snip-20.md:78]: they wouldn't enter the mempool with r == 0

File: SNIPS/snip-20.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+1. Balance checks:
+    * User token balance should cover their bid given their specified max rate $$\mathtt{balance(user)}\geq \mathtt{max\_ amount}\times r(\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$ If success, CONTINUE; else, REJECT.
+    * Paymaster STRK balance should cover user bid $$\mathtt{balance(paymaster)}\geq \mathtt{max\_ amount}\times (\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$ If success, CONTINUE; else, REJECT.
+2. Run account `__validate__`. If success CONTINUE, else REJECT.
+3. Run paymaster `__validate_paymaster__`. If success, CONTINUE; else, REVERT and skip to fee invocations below.
+4. Run account `__execute__`. If success, CONTINUE; else, REVERT and skip to fee invocations below.
+5. Fee invocations:
+    * If failure was in `__validate_paymaster__`, user pays sequencer in designated fee token:
+        * Burn $\mathtt{used\_ amount}\times r\cdot\mathtt{base\_ price}$.
```
Comment (ilyalesokhin-starkware) [inline SNIPS/snip-20.md:77]: Are we burning a non-native token in this case?

File: SNIPS/snip-20.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        * Transfer $\mathtt{used\_ amount}\times r\cdot\mathtt{tip}$.
+    * If failure was later:
+        1. Paymaster pays transaction fee to sequencer:
```
Comment (ilyalesokhin-starkware) [inline SNIPS/snip-20.md:80]: What happens if the paymaster does not have the required balance?

File: SNIPS/snip-20.md (hunk 2 above)
Comment (iliav-starkware) [reply SNIPS/snip-20.md:77]: yep

File: SNIPS/snip-20.md (hunk 1 above)
Comment (iliav-starkware) [reply SNIPS/snip-20.md:78]: r=0 will be rejected in the balance checks

File: SNIPS/snip-20.md (hunk 3 above)
Comment (leo-starkware) [reply SNIPS/snip-20.md:80]: The paymaster had the required balance at the beginning, and it seems reasonable to expect this to hold at the end of the tx if the paymaster is honest. So a solution is simply to revert the tx, and do the fee invocations on the used amount (from user and from PM, which were both holding enough).

Comment (ilyalesokhin-starkware) [reply SNIPS/snip-20.md:80]: Don't you think the most reasonable paymaster implementation is to do the token conventions in the transactions?

Do you expect the paymasters to hold a large STARK balance and do periodic token conversions?

File: SNIPS/snip-20.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+2. User submitted rate should be at least client's locally configured rate. $$\mathtt{min\_ STRK/TOKEN\_ rate}\leq r$$
+3. Balance checks:
+    * User token balance should cover their bid given their specified max exchange rate $$\mathtt{balance(user)}\geq \mathtt{max\_ amount}\times r(\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$
+    * Paymaster STRK balance should cover user bid $$\mathtt{balance(paymaster)}\geq \mathtt{max\_ amount}\times (\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$
+
+#### Block builder
+
+A paymaster contract should be used only by those who trust its code. Such users know the paymaster's exchange policies, and therefore know which rates they should submit to pass, and which rates are too frugal and may cause `__validate_paymaster__` to fail. Hence, we consider it justified to "blame" users for failed `__validate_paymaster__` and to consequently charge them. Now arises a problematic scenario: the user may hold zero STRK and the paymaster rejected the request to cover its transaction fee. In this particular case we allow the sequencer to charge the transaction fee directly in the user's designated token.
```
Comment (ilyalesokhin-starkware) [inline SNIPS/snip-20.md:63]: I would expect the exchange rate of the sequencers to be higher than the exchange rate of the paymasters.

File: SNIPS/snip-20.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+Behind the scenes, SDKs should also support sending paymasters transactions.
+
+## Rationale
+
+We submit that the above design facilitates simple and safe paymaster contracts providing the desired functionality of letting users pay transaction fees with arbitrary ERC-20 tokens.
+
+## Drawbacks
+
+1. A user who set their max rate in the range $$\mathtt{min\_ STRK/TOKEN\_ rate} \leq r\leq \mathtt{min\_ STRK/TOKEN\_ rate_accepted_by_paymaster}$$ will experience reverted `__validate_paymaster__` and pay for the reversion. As noted above, this is easily remedied by submitting high rates, especially for paymasters who charge at discounted rates.
```
Comment (ilyalesokhin-starkware) [inline SNIPS/snip-20.md:104]: The formula is broken:

"Double subscripts: use braces to clarify"

File: SNIPS/snip-20.md (hunk 4 above)
Comment (iliav-starkware) [reply SNIPS/snip-20.md:63]: Sequencers will set a low threshold just to filter overvaluations. If they set their rates too high, they will effectively block out paymaster transactions. What are you concerned about?

Comment (ilyalesokhin-starkware) [reply SNIPS/snip-20.md:63]: If the sequencer gives me a better conversion rate than the paymasters.
I would move my logic to validate, use a paymaster that always revets, and pay the sequencer with my desired token.

File: SNIPS/snip-20.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    * If failure was later:
+        1. Paymaster pays transaction fee to sequencer:
+            * Burn `used_amount × base_price`.
+            * Transfer `used_amount × tip`.
+        2. User pays paymaster: transfer `used_amount × r × (base_price(current_block)+tip)`.
+    * In case of failure during fee invocations, REVERT all state changes and return to it directly (skipping `__validate_paymaster__` and `execute`). This ensures both user and paymaster have sufficient balances due to both balance checks.
+
+### Optimization: discounted paymaster rates
+
+In the above flow, the paymaster always charges at the user's max rate `r`. This encourages strategic "bidding" on the rate. It's better to allow paymaster contracts to charge users at a smaller rate than their submitted max rate. To this end, we allow `__validate_paymaster__` to output a `charging_rate` satisfying `charging\_ rate ≤ r`. If `__validate_paymaster__` executes successfully, the fee invocation will proceed with the user's max rate replaced by the paymaster's `charging_rate`.
```
Comment (leo-starkware) [inline SNIPS/snip-20.md:93]: typo: there's still \_

Comment (leo-starkware) [reply SNIPS/snip-20.md:93]: \ _

**Review by leo-starkware — APPROVED** [review]

File: .vscode/settings.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (ilyalesokhin-starkware) [inline .vscode/settings.json]: I think this should be a separate PR.

File: SNIPS/snip-20.md (hunk 4 above)
Comment (ilyalesokhin-starkware) [reply SNIPS/snip-20.md:68]: "Sequencers will set a low threshold just to filter overvaluations." does not seem like a rational behavior.

I think Sequencers will set high threshold and actually paymaster will use r/2 or something like that.
maybe we want to do that at the protocol layer?
